API: Create/edit updates from Koji tags

nphilipp opened the pull request:
This PR contains code to let users submit new updates or edit existing ones for a Koji (side) tag, i.e. it would pull the latest builds in the tag into the update and remember the tag with the update. For editing (not fully implemented yet), it would work like this: if the list of submitted builds is non-empty, do things just as before, if it is empty, consult Koji again for a list of builds.

NB: it uses the first two commits from PR#3168, so it would be good if this one would be merged first.

nphilipp commented:
NB: This is based off of #3232 and shares the first two commits with #3168. I'd consider #3232 to be blocking it, but not the latter—it doesn't matter in which order they go in, I'll rebase the respective other accordingly and update its migration script.

nphilipp commented:
To bring some order into my PRs, this one is now based off of #3168, to avoid having to double-check common commits and migration script ordering between them.

cverna commented on bodhi/server/validators.py:
```diff
@@ -315,7 +315,7 @@ def validate_build_tags(request, **kwargs):
         valid_tags = tag_types['candidate']
 
-    for build in request.validated.get('builds', []):
+    for build in request.validated.get('builds') or []:
```
Why this change ?

nphilipp replied:
Good question, I'll remove it.

bowlofeggs replied:
```suggestion
    for build in request.validated.get('builds', []):
```

cverna commented:
@nphilipp You will probably want to squash your commits :smile:

cverna commented on bodhi/server/migrations/versions/d3f8bd499ecd_add_from_tag_column_to_updates.py:
```diff
@@ -0,0 +1,41 @@
+
+def upgrade():
+    """Add the from_tag column to the updates table."""
+    op.add_column('updates', sa.Column('from_tag', sa.UnicodeText(), nullable=True))
```
Would it be more explicit to call this `from_sidetag` ? Just plain `from_tag` might let people think that we are talking about releases tags (f30-updates, f30-updates-testing, ... ).

nphilipp replied:
I'm a very "from the bottom up" person I guess and called it `from_tag` and not `from_sidetag` because the code doesn't care if the tag you supply is a side tag or not—and AIUI can't distinguish between them either: it's just a matter of convention that we call some tags "side tags" because they're created ad-hoc and short-lived for the purpose of building packages "on the side". If you want the purpose of this API be better or explicitly reflected in the name of the flag, I'll happily make the change.

cverna replied:
I don't feel strongly on that one to be honest, so let's keep it as it is :+1:

cverna commented on bodhi/server/models.py:
```diff
@@ -121,7 +121,7 @@ def __json__(self, request=None):
 
         Args:
-            request (pyramid.util.Request): The current request.
+            request (pyramid.request.Request): The current request.
```
So these changes should be kept in separated commit from the side tag API.

nphilipp replied:
Agreed.

cverna commented on bodhi/server/models.py:
```diff
@@ -1602,6 +1602,8 @@ class Update(Base):
             Greenwave integration was not enabled when the update was created.
         compose (Compose): The :class:`Compose` that this update is currently being composed in. The
             update is locked if this is defined.
+        from_tag (unicode): The koji tag from which the list of builds was
```
```suggestion
        from_tag (str): The koji tag from which the list of builds was
```
Since Bodhi is Python 3 now the unicode type does not exists anymore :-)

nphilipp replied:
Yeah, I noticed that too while working on the CLI part of this. I'll add a tidying up commit to remove 
mentions of `unicode` and `basestr` and let the topical changes in the PR be based on that.

bowlofeggs replied:
```suggestion
        from_tag (str): The koji tag from which the list of builds was
```

cverna commented on bodhi/server/validators.py:
```diff
@@ -191,15 +191,18 @@ def cache_nvrs(request, build):
         kwargs (dict): The kwargs of the related service definition. Unused.
     """
-    for build in request.validated.get('builds', []):
+    builds = request.validated.get('builds')
```
I think this should also be in a separated commit, since it does not relate with the side tag change.

nphilipp replied:
I'm trying to remember why I changed this in the beginning... Perhaps I should remove it because I don't know what the purpose of this change was anymore.

nphilipp replied:
I've looked into it and will remove it.

cverna commented on bodhi/server/validators.py:
```diff
@@ -214,21 +217,47 @@ def validate_nvrs(request, **kwargs):
-    if not request.validated.get('builds', []):
-        request.errors.add('body', 'builds', "You may not specify an empty list of builds.")
-        return
+    builds = request.validated.get('builds') or []  # cope with builds set to None
```
I thinks these changes should also be in a separated commit from the side tag one.

nphilipp replied:
It may not look like it but this one is directly related to the from-sidetag feature, it moves checking if the list of builds is empty to the (as of yet comment-less) `else` branch in `validate_from_tag()` below.

bowlofeggs replied:
This would be more idiomatic:

```suggestion
    builds = request.validated.get('builds', [])
```

cverna commented on bodhi/server/validators.py:
```diff
@@ -1232,3 +1261,46 @@ def validate_request(request, **kwargs):
+                if "invalid taginfo" in str(e).lower():
+                    request.errors.add('body', 'from_tag', "The supplied from_tag doesn't exist.")
+                else:
+                    raise
```
Should we raise a BodhiException here ?

nphilipp replied:
Sounds like a good idea.

nphilipp replied:
Probably like this because we're good Python 3 citizens :wink::

```
    raise BodhiException(...) from e
```

cverna commented on bodhi/server/validators.py:
```diff
@@ -1232,3 +1261,46 @@ def validate_request(request, **kwargs):
+                    request.errors.add('body', 'from_tag', "The supplied from_tag doesn't exist.")
+                else:
+                    raise
+            else:
```
This is a bit hard to follow everything here :smile:, adding some comment might be a good idea to explain the logic.

nphilipp replied:
Mhm, `...  # normal update with explicit list of builds`.

cverna replied:
Yes I think that would make the code easier to read

nphilipp replied:
Heh, noticed that my proposed comment is wrong even, after looking at the greater context. I'll just move the code out of the superfluous `else:` branch (which is only necessary if the block should be handled by a `finally:` branch which we don't have here).

nphilipp commented:
> @nphilipp You will probably want to squash your commits

Of course -- do you think the first one "Use less ambiguous validator names." should be separate or part of the squashed feature?

nphilipp commented:
@cverna This latest push should address your requests. Mind that the order GitHub lists the commits is wrong (why ever), the first three are:

```
bef23ece Correct location of request type.
53cffcf0 Don't mention obsolete string types.
6af83d8f Use less ambiguous validator names.
```

Again, I'm not sure if the last of these shouldn't be squashed with the feature commits. What do you think?

abompard commented on bodhi/tests/server/services/test_updates.py:
```diff
@@ -275,6 +277,43 @@ def test_new_rpm_update_koji_error(self, *args):
+        # We don't want the new update to obsolete the existing one.
+        self.db.delete(Update.query.one())
+
+        update = self.get_update(builds=None, from_tag='f17-updates-candidate')
```
This fails in CI because the code does not expect builds to be `None`, only a list (that can be empty). I suggest to either use `[]` here or to handle in the validation code the case where `builds` is `None`.

nphilipp replied:
Where does it fail and how? AIUI, it uses `BaseTestCaseMixin.get_update()` which has this:

```
        if builds:
            if isinstance(builds, list):
                builds = ','.join(builds)
            if not isinstance(builds, str):
                builds = builds.encode('utf-8')
            update['builds'] = builds
```

I.e. if `builds` is set to `None`, it should simply be missing from the returned dictionary.

abompard replied:
In the CI logs you'll see:
```
13:55:12  Traceback (most recent call last):
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/tweens.py", line 41, in excview_tween
13:55:12      response = handler(request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/router.py", line 148, in handle_request
13:55:12      registry, request, context, context_iface, view_name
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/view.py", line 667, in _call_view
13:55:12      response = view_callable(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/config/views.py", line 169, in __call__
13:55:12      return view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/config/views.py", line 188, in attr_view
13:55:12      return view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
13:55:12      return view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/viewderivers.py", line 384, in authdebug_view
13:55:12      return view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/viewderivers.py", line 325, in secured_view
13:55:12      return view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
13:55:12      result = view(context, request)
13:55:12    File "/usr/lib/python3.7/site-packages/pyramid/viewderivers.py", line 144, in _requestonly_view
13:55:12      response = view(request)
13:55:12    File "/usr/lib/python3.7/site-packages/cornice/service.py", line 493, in wrapper
13:55:12      validator(request, **args)
13:55:12    File "/bodhi/bodhi/server/validators.py", line 90, in validator
13:55:12      f(request, **kwargs)
13:55:12    File "/bodhi/bodhi/server/validators.py", line 203, in validate_build_nvrs
13:55:12      for build in request.validated.get('builds', []):
13:55:12  TypeError: 'NoneType' object is not iterable
```
The `builds` key is set to `None`, so `get()` does not fallback on `[]`, therefore the crash.

abompard commented:
An improper testcase crashes CI (I've suggested possible fixes in a line comment), other than that the code looks good, thanks :-)

cverna commented:
> @cverna This latest push should address your requests. Mind that the order GitHub lists the commits is wrong (why ever), the first three are:
> 
> ```
> bef23ece Correct location of request type.
> 53cffcf0 Don't mention obsolete string types.
> 6af83d8f Use less ambiguous validator names.
> ```
> 
> Again, I'm not sure if the last of these shouldn't be squashed with the feature commits. What do you think?

I think renaming the validators can be squashed with the feature commit since it is kind of related.

nphilipp commented:
@cverna > I think renaming the validators can be squashed with the feature commit since it is kind of related.

Will do.

bowlofeggs commented:
Can you move the "Don't mention obsolete string types." commit to another pull request, since it's unrelated to the other changes here? Alternatively, please squash the other commits to one and then this can stay here.

bowlofeggs commented:
```flake8``` is upset about some lines in this PR:

```
11:08:37  ./bodhi/tests/server/test_validators.py:789:36: W504 line break after binary operator
11:08:37          assert (str(excinfo.value) ==
11:08:37                                     ^
11:08:37  ./bodhi/tests/server/test_validators.py:790:1: W191 indentation contains tabs
11:08:37  				"Encountered error while requesting tagged builds from Koji: 'foo'")
11:08:37  ^
11:08:37  ./bodhi/tests/server/test_validators.py:790:1: E101 indentation contains mixed spaces and tabs
11:08:37  				"Encountered error while requesting tagged builds from Koji: 'foo'")
11:08:37  ^
11:08:37  ./bodhi/tests/server/test_validators.py:790:5: E128 continuation line under-indented for visual indent
11:08:37  				"Encountered error while requesting tagged builds from Koji: 'foo'")
11:08:37      ^
11:08:37  ./bodhi/tests/server/test_validators.py:792:1: E101 indentation contains mixed spaces and tabs
11:08:37          # check type of wrapped exception
11:08:37  ^
```

bowlofeggs commented on bodhi/server/validators.py:
```diff
@@ -191,15 +192,15 @@ def cache_nvrs(request, build):
         kwargs (dict): The kwargs of the related service definition. Unused.
     """
-    for build in request.validated.get('builds', []):
+    for build in request.validated.get('builds') or []:  # cope with builds being None
```
This would be more idiomatic:

```suggestion
    for build in request.validated.get('builds', []):
```

nphilipp replied:
But it wouldn't be the same, that's why I replaced it (I want it to not croak at the `'builds'` key having `None` as a value). I could pull it apart like this to make it stand out more:

```
builds = request.validated.get('builds')
# cope with builds being None
if not builds:
    builds = []
for build in builds:
    ...
```

nphilipp replied:
@bowlofeggs I didn't make it clear this was a question: should I change this as I outlined, or add a comment?

bowlofeggs commented on bodhi/server/validators.py:
```diff
@@ -496,16 +523,16 @@ def validate_acls(request, **kwargs):
 
 @postschema_validator
-def validate_uniqueness(request, **kwargs):
+def validate_build_uniqueness(request, **kwargs):
```
I'm fine with renaming this, but it seems unrelated to the rest of the PR - can we do this in a separate pull request?

nphilipp replied:
Sure, I'll put this and the two type-change ones in their own PR (one PR for the three commits).

bowlofeggs commented:
It would also be ideal to move the commit "Correct location of request type." to its own PR. It would make this one simpler to review.

nphilipp commented:
I'll move the standalone commits out into their own PR.

nphilipp commented:
The first three commits are covered in their own PR #3320.

nphilipp commented:
Apparently GH doesn't let me resolve this without dismissing. :disappointed:

cverna commented on bodhi/server/migrations/versions/d3f8bd499ecd_add_from_tag_column_to_updates.py:
```diff
@@ -0,0 +1,41 @@
+
+# revision identifiers, used by Alembic.
+revision = 'd3f8bd499ecd'
+down_revision = '5703ddfe855d'
```
You need to update the migration script to use the latest revision :smile:

abompard replied:
Yep, this should be rebased.

cverna commented:
Overall this LGTM. I think you can squash your commits. 

@abompard, do you want to also take a second look at it ?

nphilipp committed:
API: Create/edit updates from Koji tags.

This lets users populate builds from Koji (side-)tags when creating
updates (or repopulate when editing them). it also stores the tag in
question in Update.from_tag.

Squashed commit of the following:

commit c47e2dfbe55a2a0984ead8e1ec6a5ff6f3fee02a
Author: Nils Philippsen <nils@redhat.com>
Date:   Fri May 17 18:33:56 2019 +0200

    Move Koji tag expansion to validate_from_tag().

    This finds the latest builds in a tag in the from_tag validator and adds
    them to request.validated. This is necessary because the list of builds
    is further used in validate_acls() to decide if a user is allowed to
    submit an update for the involved builds.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit df05cebf8aec9b3c7ca20c350ea2acef44c1e4f7
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 15:24:54 2019 +0200

    Create updates from a Koji tag.

    This lets users specify a Koji tag instead of an explicit set of builds,
    from which the latest component builds will be pulled.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit 665aa086009789da052c2721a82a795466a7ac7b
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 15:22:37 2019 +0200

    Add from_tag parameter to get_updates().

    This allows testing updates where the builds should be filled from all
    latest builds in a Koji tag.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit c84a637c70753a6bc333ba0dcb62d98a39336d58
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 15:16:10 2019 +0200

    Add Update.from_tag.

    This column stores the tag from which the set of builds of an update was
    filled initially.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit 0c0717d374f6f6bbcb51e60161576c1a389ca325
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 12:22:42 2019 +0200

    Accept from_tag alternatively to builds.

    This is to let users create updates from all latest builds of a Koji
    tag.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit d6bf4fa6b69186bee4e4b01766d47efb78220dae
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 11:05:23 2019 +0200

    Add validate_builds_or_from_tag_exist().

    This validator checks that at minimum one of `builds` or `from_tag` is
    supplied as parameters and isn't empty.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit e5e3480901cb7d9e90cb3a986c667790951e37fb
Author: Nils Philippsen <nils@redhat.com>
Date:   Tue May 14 10:58:47 2019 +0200

    Make DevBuildsys.listTagged() honor `latest` keyword.

    Previously, it could return multiple builds of the same component in a
    tag. This required updating tests which expected these wrong results.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

commit 68953597b51297c7ca33ae18976018574bedec6f
Author: Nils Philippsen <nils@redhat.com>
Date:   Fri May 10 12:08:19 2019 +0200

    Add validate_from_tag().

    This validator ensures that if from_tag is supplied in the request, it
    is a valid Koji tag.

    Signed-off-by: Nils Philippsen <nils@redhat.com>

Fixes: #3009

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
The latest push squashes the commits into one.

nphilipp commented:
Oops, missed that another migration went in meanwhile...

abompard reviewed (approved):
I think this one looks good now.

mergify[bot] committed:
Merge branch 'develop' into develop--update-from-sidetag-api

cverna reviewed (approved):
:fireworks:

mergify[bot] committed:
Merge branch 'develop' into develop--update-from-sidetag-api
